source/dofs/dof_renumbering.cc

```diff
@@ -2819,72 +2819,71 @@ namespace DoFRenumbering
-    //     within the other category.
+    //     dof is touched -> touch_count
+    // (c) determine by how many MPI processes a dof is touched
+    //     -> dofs_by_rank_access
+    // (d) combine both category types of (b) and (c) and list the indices
+    //     according to four categories
 
     const std::vector<std::vector<unsigned int>> dofs_by_rank_access =
       group_dofs_by_rank_access(
         *matrix_free.get_dof_info(component).vector_partitioner);
 
-    const std::pair<std::vector<unsigned int>, std::vector<unsigned char>>
-      local_numbering = compute_mf_numbering(matrix_free, component);
+    const auto &[local_numbering, touch_count] =
+      compute_mf_numbering(matrix_free, component);
 
-    // Now construct the new numbering
     const IndexSet &owned_dofs = matrix_free.get_dof_info(component)
                                    .vector_partitioner->locally_owned_range();
-    std::vector<unsigned int> new_numbers;
-    new_numbers.reserve(owned_dofs.n_elements());
-
-    // step 1: Take all DoFs with reference only to the current MPI process
-    // and touched once ("perfect overlap" case). We define a custom
-    // comparator for std::sort to then order the unknowns by the specified
-    // matrix-free loop order
-    const std::vector<unsigned int> &purely_local_dofs = dofs_by_rank_access[0];
-    for (const unsigned int i : purely_local_dofs)
-      if (local_numbering.second[i] == 1)
-        new_numbers.push_back(i);
-
-    const auto comparator = [&](const unsigned int a, const unsigned int b) {
-      return (local_numbering.first[a] < local_numbering.first[b]);
-    };
-
-    std::sort(new_numbers.begin(), new_numbers.end(), comparator);
-    unsigned int sorted_size = new_numbers.size();
-
-    // step 2: Take all DoFs with reference to only the current MPI process
-    // and touched multiple times (more strain on caches).
-    for (auto i : purely_local_dofs)
-      if (local_numbering.second[i] > 1)
-        new_numbers.push_back(i);
-    std::sort(new_numbers.begin() + sorted_size, new_numbers.end(), comparator);
-    sorted_size = new_numbers.size();
-
-    // step 3: Get all DoFs with reference from other MPI ranks
+    const types::global_dof_index locally_owned_size = owned_dofs.n_elements();
+    AssertDimension(locally_owned_size, local_numbering.size());
+    AssertDimension(locally_owned_size, touch_count.size());
+
+    // Create a second permutation to group the unknowns into the following
+    // four categories, to be eventually composed with 'local_renumbering':
+    // 0: all DoFs without any reference (constrained DoFs)
+    // 1: all DoFs with reference only to the current MPI process and
+    //    touched once ("perfect overlap" case)
+    // 2: all DoFs with reference to only the current MPI process and
+    //    touched from multiple cell ranges (more strain on caches)
+    // 3: all DoFs with reference from other MPI ranks
```

Comment:
Would it work if you'd create an `enum` for this distinction? I believe that this would make the code easier to read. What do you think about it?

Reply:
That's a good idea, I can do it later today.

Reply:
I applied an enum to get better names.